copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+					result.Error = err.Error()
+					continue
+				}
+				// record where it points, both by itself (it
+				// could be a relative link) and in the context
+				// of the chroot
+				result.ImmediateTarget = immediateTarget
+				resolvedTarget, err := resolvePath(request.Root, globbed, pm)
+				if err != nil {
+					return errorResponse("copier: stat: error resolving %q: %v", globbed, err)
+				}
+				// lstat the thing that we point to
+				info, err := os.Lstat(resolvedTarget)
+				if err != nil {
+					result.Error = err.Error()
+					continue
+				}
+				// replace IsArchive/IsDir/IsRegular with info about the target
+				if info.Mode().IsRegular() && request.StatOptions.CheckForArchives {
+					result.IsArchive = isArchivePath(resolvedTarget)
+					checkForArchive = false
+				}
+				result.IsDir = info.IsDir()
+				result.IsRegular = info.Mode().IsRegular()
+			}
+			if result.IsRegular && checkForArchive {
+				// we were asked to check on this, and it
+				// wasn't a symlink, in which case we'd have
+				// already checked what the link points to
+				result.IsArchive = isArchivePath(globbed)
+			}
+		}
+		// no unskipped matches -> error
+		if len(s.Globbed) == 0 {
+			s.Globbed = nil
+			s.Results = nil
+			s.Error = fmt.Sprintf("copier: stat: %q: %v", glob, syscall.ENOENT)
+		}
+	}
+	return &Response{Stat: StatResponse{Globs: stats}}
+}
+
+func copierHandlerGet(bulkWriter io.Writer, request Request, pm *fileutils.PatternMatcher, idMappings *idtools.IDMappings) (*Response, func() error, error) {
+	statRequest := request
+	statRequest.Request = RequestStat
+	statResponse := copierHandlerStat(request, pm)
+	errorResponse := func(fmtspec string, args ...interface{}) (*Response, func() error, error) {
+		return &Response{Stat: statResponse.Stat, Get: GetResponse{Error: fmt.Sprintf(fmtspec, args...)}}, nil, nil
+	}
+	if statResponse.Stat.Error != "" {
+		return errorResponse("%s", statResponse.Stat.Error)
+	}
+	if len(request.Globs) == 0 {
+		return errorResponse("copier: get: expected at least one glob pattern, got 0")
+	}
+	// build a queue of items by globbing
+	var queue []string
+	globMatchedCount := 0
+	for _, glob := range request.Globs {
+		globMatched, err := filepath.Glob(glob)
+		if err != nil {
+			return errorResponse("copier: get: glob %q: %v", glob, err)
+		}
+		globMatchedCount += len(globMatched)
+		filtered := make([]string, 0, len(globMatched))
+		for _, globbed := range globMatched {
+			rel, excluded, err := pathIsExcluded(request.Root, globbed, pm)
+			if err != nil {
+				return errorResponse("copier: get: checking if %q is excluded: %v", globbed, err)
+			}
+			if rel == "." || !excluded {
+				filtered = append(filtered, globbed)
+			}
+		}
+		if len(filtered) == 0 {
+			return errorResponse("copier: get: glob %q matched nothing (%d filtered out of %v): %v", glob, len(globMatched), globMatched, syscall.ENOENT)
+		}
+		queue = append(queue, filtered...)
+	}
+	// no matches -> error
+	if len(queue) == 0 {
```

Comment:
(This is unreachable AFAICS, the code already checks for at least one glob and at least one `filtered` result.)